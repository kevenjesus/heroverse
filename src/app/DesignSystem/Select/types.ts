export interface OptionType {
    label: string
    value: string | number
}

export interface SelectProps {
    value?: string | number | null | undefined
    placeholder?: string
    onChange: (item: OptionType) => void
    options: OptionType[]
}