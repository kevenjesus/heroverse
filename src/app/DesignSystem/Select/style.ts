import { breakpoints, colors } from '@/app/styles/guide'
import styled from  'styled-components'

export const Wrapper = styled.div`
    position: relative;
`

export const Overlay = styled.div`
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    position: fixed;
    z-index: 5;
    background: rgba(0,0,0,0.7);
    @media (min-width: ${breakpoints.lg}) {
        display: none;
    }
`

export const SelectField = styled.div`
    width: 100%;
    height: 60px;
    background: #fff;
    border: 2px solid ${colors.black};
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 0 20px;
    cursor: pointer;
`

export const SelectBox = styled.div`
    position: fixed;
    width: 100%;
    bottom: 0;
    left: 0;
    z-index: 6;
    padding: 20px;
    background: #fff;
    @media (min-width: ${breakpoints.lg}) {
        position: absolute;
        top: 50px;
        padding: 0;
    }
`

export const SelectBoxContent = styled.div`
    height: 300px;
    display: flex;
    flex-direction: column;
    overflow-y: scroll;
    background-color: #fff;
    &::-webkit-scrollbar {
        width: 10px;
    }
    &::-webkit-scrollbar-thumb {
        background: rgba(62, 56, 56, 0.2);
    }
    @media (min-width: ${breakpoints.lg}) {
        height: auto;
        overflow-y: initial;
        border: 2px solid ${colors.black};
    }
`

export const SelectItem = styled.div`
    width: 100%;
    height: 60px;
    background: #fff
    display: flex;
    align-items: center;
    padding: 16px;
    cursor: pointer;
    &:hover {
        background: rgba(142, 202, 230, 0.2);;
    }
`
