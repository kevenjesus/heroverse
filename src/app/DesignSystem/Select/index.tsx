import { useCallback, useEffect, useRef, useState } from 'react'
import Typographic from '../Typographic'
import * as S from './style'
import { OptionType, SelectProps } from './types'

export default function Select({options, value, placeholder = 'Selecione', onChange}: SelectProps) {
    const [isOpen, setOpen] = useState(false)
    const wrapperRef = useRef<HTMLDivElement>(null)

    const handleChange = useCallback((item: OptionType) => {
        onChange(item)
        setOpen(false)
    }, [])

    useEffect(() => {
        const handleMouseDown = (event: MouseEvent) => {
            if (wrapperRef.current && !wrapperRef.current.contains(event.target as Node)) {
                setOpen(false)
            }
        }
        document.addEventListener('mousedown', handleMouseDown)
        return () => {
            document.removeEventListener('mousedown', handleMouseDown)
        }
    }, [])
    return (
        <S.Wrapper ref={wrapperRef}>
            <S.SelectField onClick={() => setOpen(!isOpen)}>
                <Typographic>{value ? value : placeholder}</Typographic>
                <img width={13} height={13} src='/arrow-down.png' alt="Selecionar" />
            </S.SelectField>
            {
                isOpen && (
                <>
                    <S.Overlay role="banner" aria-label='overlay' onClick={() => setOpen(false)} />
                    <S.SelectBox>
                        <S.SelectBoxContent>
                            {
                                options.map(option => (
                                    <S.SelectItem onClick={() => handleChange(option)} key={option.value}>{option.label}</S.SelectItem>
                                ))
                            }
                        </S.SelectBoxContent>
                    </S.SelectBox>
                </>
                )
            }
        </S.Wrapper>
    )
}