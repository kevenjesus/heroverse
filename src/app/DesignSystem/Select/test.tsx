import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Select from '.';

const options = [
  { label: 'Option 1', value: 1 },
  { label: 'Option 2', value: 2 },
  { label: 'Option 3', value: 3 },
];

describe('Select', () => {
  it('should display placeholder text when no option is selected', () => {
    const { getByText } = render(<Select options={options} onChange={() => {}} />);
    expect(getByText('Selecione')).toBeInTheDocument();
  });

  it('should display selected option', () => {
    const { getByText } = render(
      <Select options={options} value='Option 2' onChange={() => {}} />
    );
    expect(getByText('Option 2')).toBeInTheDocument();
  });

  it('should display options when select field is clicked', () => {
    const { getByText } = render(<Select options={options} onChange={() => {}} />);
    fireEvent.click(getByText('Selecione'));
    expect(getByText('Option 1')).toBeInTheDocument();
    expect(getByText('Option 2')).toBeInTheDocument();
    expect(getByText('Option 3')).toBeInTheDocument();
  });

  it('should call onChange when an option is clicked', () => {
    const handleChange = jest.fn();
    const { getByText } = render(<Select options={options} onChange={handleChange} />);
    fireEvent.click(getByText('Selecione'));
    fireEvent.click(getByText('Option 1'));
    expect(handleChange).toHaveBeenCalledWith({ label: 'Option 1', value: 1 });
  });

  it('should close overlay when clicking on overlay', async () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 800 });
    const { getByText } = render(<Select options={[{ label: 'Option 1', value: '1' }]} value={null} onChange={() => {}} />);
    fireEvent.click(getByText('Selecione'));
    const overlay = await screen.findByRole('banner', { name: 'overlay' });
    fireEvent.click(overlay);
    expect(overlay).not.toBeInTheDocument();
  });
  

  it('should close select when clicking outside', () => {
    const { container } = render(<Select options={[{ label: 'Option 1', value: '1' }]} value={null} onChange={() => {}} />);
    const outsideElement = document.createElement('div');
    document.body.appendChild(outsideElement);
    fireEvent.mouseDown(outsideElement);
    expect(container.firstChild).not.toHaveClass('open');
  });
  
});
