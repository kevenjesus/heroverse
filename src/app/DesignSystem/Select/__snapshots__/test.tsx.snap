// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should display placeholder text when no option is selected 1`] = `
<div>
  <div
    class="sc-hLseeU dcExOP"
  >
    <div
      class="sc-jTrPJq kBePcy"
    >
      <p
        class="sc-dmqHEX hZDiJS"
      >
        Selecione
      </p>
      <img
        alt="Selecionar"
        height="13"
        src="/arrow-down.png"
        width="13"
      />
    </div>
  </div>
</div>
`;
